metaflow/includefile.py

```diff
@@ -70,7 +248,107 @@ def __repr__(self):
+        return_value = {'type': url_type, 'url': url}
+        return_value.update(kwargs)
+        return json.dumps(return_value)
+
+    @staticmethod
+    def decode_value(value):
+        if value is None or len(value) == 0 or value[0] != '{':
+            return {'type': 'base', 'url': value}
+        return json.loads(value)
+
+    def store(self, flow_name, path, is_text, encoding, logger):
+        sz = os.path.getsize(path)
+        unit = ['B', 'KB', 'MB', 'GB', 'TB']
+        pos = 0
+        while pos < len(unit) and sz >= 1024:
+            sz = sz // 1024
+            pos += 1
+        if pos >= 3:
+            extra = '(this may take a while)'
+        else:
+            extra = ''
+        logger(
+            'Including file %s of size %d%s %s' % (path, sz, unit[pos], extra))
+        try:
+            cur_obj = TransformableObject(io.open(path, mode='rb').read())
+        except IOError:
+            # If we get an error here, since we know that the file exists already,
+            # it means that read failed which happens with Python 2.7 for large files
+            raise MetaflowException('Cannot read file at %s -- this is likely because it is too '
+                                    'large to be properly handled by Python 2.7' % path)
+        sha = sha1(cur_obj.current()).hexdigest()
+        path = os.path.join(self._client_class.get_root_from_config(logger, True), flow_name, sha)
+        buf = BytesIO()
+        with gzip.GzipFile(
+                fileobj=buf, mode='wb', compresslevel=3) as f:
+            f.write(cur_obj.current())
+        cur_obj.transform(lambda _: buf)
+        buf.seek(0)
+        with self._client_class() as client:
+            url = client.put(path, buf.getvalue(), overwrite=False)
+            logger('File persisted at %s' % url)
+            return Uploader.encode_url(Uploader.file_type, url, is_text=is_text, encoding=encoding)
+
+    def load(self, value):
+        value_info = Uploader.decode_value(value)
+        with self._client_class() as client:
+            obj = client.get(value_info['url'], return_missing=True)
+            if obj.exists:
+                if value_info['type'] == Uploader.file_type:
+                    # We saved this file directly so we know how to read it out
+                    with gzip.GzipFile(filename=obj.path, mode='rb') as f:
+                        if value_info['is_text']:
+                            return io.TextIOWrapper(f, encoding=value_info.get('encoding')).read()
+                        return f.read()
+                else:
+                    # We open this file according to the is_text and encoding information
+                    if value_info['is_text']:
+                        return io.open(obj.path, mode='rt', encoding=value_info.get('encoding')).read()
+                    else:
+                        return io.open(obj.path, mode='rb').read()
+            raise FileNotFoundError("File at %s does not exist" % self._url)
```

Comment:
Oops, reading here, this is an error; it should be value_info['url'] and not self._url. I will fix this.